File: sdks/typescript/packages/mechanisms/polkadot/test/unit/constants.test.ts
import { describe, it, expect } from "vitest";
import {
  POLKADOT_CAIP2_NAMESPACE,
  POLKADOT_ASSET_HUB_CAIP2,
  KUSAMA_ASSET_HUB_CAIP2,
  WESTEND_ASSET_HUB_CAIP2,
  SCHEME_EXACT_DIRECT,
  DEFAULT_POLKADOT_INDEXER,
  DEFAULT_KUSAMA_INDEXER,
  DEFAULT_WESTEND_INDEXER,
  DEFAULT_POLKADOT_RPC,
  DEFAULT_KUSAMA_RPC,
  DEFAULT_WESTEND_RPC,
  POLKADOT_NETWORKS,
  getNetworkConfig,
  isPolkadotNetwork,
} from "../../src/constants";

describe("Polkadot Constants", () => {
  describe("CAIP-2 Namespace", () => {
    it("should have correct namespace", () => {
      expect(POLKADOT_CAIP2_NAMESPACE).toBe("polkadot");
    });
  });

  describe("Network Identifiers", () => {
    it("should have correct Polkadot Asset Hub CAIP-2 identifier", () => {
      expect(POLKADOT_ASSET_HUB_CAIP2).toBe("polkadot:68d56f15f85d3136970ec16946040bc1");
    });

    it("should have correct Kusama Asset Hub CAIP-2 identifier", () => {
      expect(KUSAMA_ASSET_HUB_CAIP2).toBe(
        "polkadot:48239ef607d7928874027a43a67689209727dfb3d3dc5e5b03a39bdc2eda771a"
      );
    });

    it("should have correct Westend Asset Hub CAIP-2 identifier", () => {
      expect(WESTEND_ASSET_HUB_CAIP2).toBe("polkadot:e143f23803ac50e8f6f8e62695d1ce9e");
    });

    it("should start with polkadot: prefix", () => {
      expect(POLKADOT_ASSET_HUB_CAIP2.startsWith("polkadot:")).toBe(true);
      expect(KUSAMA_ASSET_HUB_CAIP2.startsWith("polkadot:")).toBe(true);
      expect(WESTEND_ASSET_HUB_CAIP2.startsWith("polkadot:")).toBe(true);
    });
  });

  describe("Scheme", () => {
    it("should have correct scheme identifier", () => {
      expect(SCHEME_EXACT_DIRECT).toBe("exact-direct");
    });
  });

  describe("Indexer URLs", () => {
    it("should have valid Subscan URLs", () => {
      expect(DEFAULT_POLKADOT_INDEXER).toBe("https://assethub-polkadot.api.subscan.io");
      expect(DEFAULT_KUSAMA_INDEXER).toBe("https://assethub-kusama.api.subscan.io");
      expect(DEFAULT_WESTEND_INDEXER).toBe("https://assethub-westend.api.subscan.io");
    });

    it("should use HTTPS", () => {
      expect(DEFAULT_POLKADOT_INDEXER.startsWith("https://")).toBe(true);
      expect(DEFAULT_KUSAMA_INDEXER.startsWith("https://")).toBe(true);
      expect(DEFAULT_WESTEND_INDEXER.startsWith("https://")).toBe(true);
    });
  });

  describe("RPC URLs", () => {
    it("should have valid WebSocket RPC URLs", () => {
      expect(DEFAULT_POLKADOT_RPC).toBe("wss://polkadot-asset-hub-rpc.polkadot.io");
      expect(DEFAULT_KUSAMA_RPC).toBe("wss://kusama-asset-hub-rpc.polkadot.io");
      expect(DEFAULT_WESTEND_RPC).toBe("wss://westend-asset-hub-rpc.polkadot.io");
    });

    it("should use WSS", () => {
      expect(DEFAULT_POLKADOT_RPC.startsWith("wss://")).toBe(true);
      expect(DEFAULT_KUSAMA_RPC.startsWith("wss://")).toBe(true);
      expect(DEFAULT_WESTEND_RPC.startsWith("wss://")).toBe(true);
    });
  });

  describe("Network Configurations", () => {
    it("should have all three networks configured", () => {
      expect(Object.keys(POLKADOT_NETWORKS)).toHaveLength(3);
      expect(POLKADOT_NETWORKS[POLKADOT_ASSET_HUB_CAIP2]).toBeDefined();
      expect(POLKADOT_NETWORKS[KUSAMA_ASSET_HUB_CAIP2]).toBeDefined();
      expect(POLKADOT_NETWORKS[WESTEND_ASSET_HUB_CAIP2]).toBeDefined();
    });

    it("should have correct Polkadot Asset Hub config", () => {
      const config = POLKADOT_NETWORKS[POLKADOT_ASSET_HUB_CAIP2];
      expect(config.name).toBe("Polkadot Asset Hub");
      expect(config.caip2).toBe(POLKADOT_ASSET_HUB_CAIP2);
      expect(config.ss58Prefix).toBe(0);
      expect(config.isTestnet).toBe(false);
      expect(config.genesisHash.startsWith("0x")).toBe(true);
    });

    it("should have correct Kusama Asset Hub config", () => {
      const config = POLKADOT_NETWORKS[KUSAMA_ASSET_HUB_CAIP2];
      expect(config.name).toBe("Kusama Asset Hub");
      expect(config.ss58Prefix).toBe(2);
      expect(config.isTestnet).toBe(false);
    });

    it("should have correct Westend Asset Hub config (testnet)", () => {
      const config = POLKADOT_NETWORKS[WESTEND_ASSET_HUB_CAIP2];
      expect(config.name).toBe("Westend Asset Hub");
      expect(config.ss58Prefix).toBe(42);
      expect(config.isTestnet).toBe(true);
    });

    it("should have all required fields in configs", () => {
      Object.values(POLKADOT_NETWORKS).forEach((config) => {
        expect(config.name).toBeDefined();
        expect(config.caip2).toBeDefined();
        expect(config.rpcUrl).toBeDefined();
        expect(config.indexerUrl).toBeDefined();
        expect(config.genesisHash).toBeDefined();
        expect(typeof config.ss58Prefix).toBe("number");
        expect(typeof config.isTestnet).toBe("boolean");
      });
    });
  });

  describe("getNetworkConfig", () => {
    it("should return config for valid networks", () => {
      const config = getNetworkConfig(POLKADOT_ASSET_HUB_CAIP2);
      expect(config).toBeDefined();
      expect(config?.name).toBe("Polkadot Asset Hub");
    });

    it("should return undefined for invalid networks", () => {
      expect(getNetworkConfig("eip155:1")).toBeUndefined();
      expect(getNetworkConfig("polkadot:unknown")).toBeUndefined();
      expect(getNetworkConfig("")).toBeUndefined();
    });
  });

  describe("isPolkadotNetwork", () => {
    it("should return true for Polkadot networks", () => {
      expect(isPolkadotNetwork(POLKADOT_ASSET_HUB_CAIP2)).toBe(true);
      expect(isPolkadotNetwork(KUSAMA_ASSET_HUB_CAIP2)).toBe(true);
      expect(isPolkadotNetwork(WESTEND_ASSET_HUB_CAIP2)).toBe(true);
      expect(isPolkadotNetwork("polkadot:anychain")).toBe(true);
    });

    it("should return false for non-Polkadot networks", () => {
      expect(isPolkadotNetwork("eip155:1")).toBe(false);
      expect(isPolkadotNetwork("ton:mainnet")).toBe(false);
      expect(isPolkadotNetwork("ethereum")).toBe(false);
      expect(isPolkadotNetwork("")).toBe(false);
    });
  });
});

File: sdks/go/mechanisms/polkadot/constants_test.go
package polkadot

import (
	"strings"
	"testing"
)

func TestSchemeConstants(t *testing.T) {
	if SchemeExactDirect != "exact-direct" {
		t.Errorf("SchemeExactDirect = %v, want exact-direct", SchemeExactDirect)
	}
}

func TestCAIP2Identifiers(t *testing.T) {
	tests := []struct {
		name     string
		caip2    string
		wantName string
	}{
		{
			name:     "Polkadot Asset Hub",
			caip2:    PolkadotAssetHubCAIP2,
			wantName: "Polkadot Asset Hub",
		},
		{
			name:     "Kusama Asset Hub",
			caip2:    KusamaAssetHubCAIP2,
			wantName: "Kusama Asset Hub",
		},
		{
			name:     "Westend Asset Hub",
			caip2:    WestendAssetHubCAIP2,
			wantName: "Westend Asset Hub",
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			if !strings.HasPrefix(tt.caip2, "polkadot:") {
				t.Errorf("CAIP-2 identifier %v should start with 'polkadot:'", tt.caip2)
			}

			config, ok := GetNetworkConfig(tt.caip2)
			if !ok {
				t.Errorf("GetNetworkConfig(%v) returned false", tt.caip2)
				return
			}
			if config.Name != tt.wantName {
				t.Errorf("GetNetworkConfig(%v).Name = %v, want %v", tt.caip2, config.Name, tt.wantName)
			}
		})
	}
}

func TestPolkadotAssetHubCAIP2Format(t *testing.T) {
	if PolkadotAssetHubCAIP2 != "polkadot:68d56f15f85d3136970ec16946040bc1" {
		t.Errorf("PolkadotAssetHubCAIP2 = %v, want polkadot:68d56f15f85d3136970ec16946040bc1", PolkadotAssetHubCAIP2)
	}
}

func TestIndexerEndpoints(t *testing.T) {
	tests := []struct {
		name    string
		url     string
		wantURL string
	}{
		{
			name:    "Polkadot Asset Hub Indexer",
			url:     PolkadotAssetHubIndexer,
			wantURL: "https://assethub-polkadot.api.subscan.io",
		},
		{
			name:    "Kusama Asset Hub Indexer",
			url:     KusamaAssetHubIndexer,
			wantURL: "https://assethub-kusama.api.subscan.io",
		},
		{
			name:    "Westend Asset Hub Indexer",
			url:     WestendAssetHubIndexer,
			wantURL: "https://assethub-westend.api.subscan.io",
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			if tt.url != tt.wantURL {
				t.Errorf("%v = %v, want %v", tt.name, tt.url, tt.wantURL)
			}
			if !strings.HasPrefix(tt.url, "https://") {
				t.Errorf("%v should use HTTPS", tt.name)
			}
		})
	}
}

func TestRPCEndpoints(t *testing.T) {
	tests := []struct {
		name string
		url  string
	}{
		{"Polkadot Asset Hub RPC", PolkadotAssetHubRPC},
		{"Kusama Asset Hub RPC", KusamaAssetHubRPC},
		{"Westend Asset Hub RPC", WestendAssetHubRPC},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			if !strings.HasPrefix(tt.url, "wss://") {
				t.Errorf("%v should use WSS protocol, got %v", tt.name, tt.url)
			}
		})
	}
}

func TestUSDTTokenInfo(t *testing.T) {
	tests := []struct {
		name  string
		token TokenInfo
	}{
		{"USDT Polkadot", USDTPolkadot},
		{"USDT Kusama", USDTKusama},
		{"USDT Westend", USDTWestend},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			// All USDT tokens should have asset ID 1984
			if tt.token.AssetID != 1984 {
				t.Errorf("%v.AssetID = %v, want 1984", tt.name, tt.token.AssetID)
			}
			// All USDT tokens should have 6 decimals
			if tt.token.Decimals != 6 {
				t.Errorf("%v.Decimals = %v, want 6", tt.name, tt.token.Decimals)
			}
			// All should have USDT symbol
			if tt.token.Symbol != "USDT" {
				t.Errorf("%v.Symbol = %v, want USDT", tt.name, tt.token.Symbol)
			}
		})
	}
}

func TestNetworkConfigHasCorrectCAIP2(t *testing.T) {
	for caip2, config := range Networks {
		t.Run(config.Name, func(t *testing.T) {
			if config.CAIP2 != caip2 {
				t.Errorf("Networks[%v].CAIP2 = %v, want %v", caip2, config.CAIP2, caip2)
			}
		})
	}
}

func TestNetworkConfigHasGenesisHash(t *testing.T) {
	for caip2, config := range Networks {
		t.Run(config.Name, func(t *testing.T) {
			if config.GenesisHash == "" {
				t.Errorf("Networks[%v].GenesisHash is empty", caip2)
			}
			if !strings.HasPrefix(config.GenesisHash, "0x") {
				t.Errorf("Networks[%v].GenesisHash should start with 0x", caip2)
			}
			// Genesis hash should be 66 characters (0x + 64 hex chars)
			if len(config.GenesisHash) != 66 {
				t.Errorf("Networks[%v].GenesisHash length = %v, want 66", caip2, len(config.GenesisHash))
			}
		})
	}
}

func TestNetworkConfigSS58Prefix(t *testing.T) {
	tests := []struct {
		caip2       string
		wantPrefix  int
		wantTestnet bool
	}{
		{PolkadotAssetHubCAIP2, 0, false},
		{KusamaAssetHubCAIP2, 2, false},
		{WestendAssetHubCAIP2, 42, true},
	}

	for _, tt := range tests {
		t.Run(tt.caip2, func(t *testing.T) {
			config, ok := GetNetworkConfig(tt.caip2)
			if !ok {
				t.Fatalf("GetNetworkConfig(%v) returned false", tt.caip2)
			}
			if config.SS58Prefix != tt.wantPrefix {
				t.Errorf("GetNetworkConfig(%v).SS58Prefix = %v, want %v", tt.caip2, config.SS58Prefix, tt.wantPrefix)
			}
			if config.IsTestnet != tt.wantTestnet {
				t.Errorf("GetNetworkConfig(%v).IsTestnet = %v, want %v", tt.caip2, config.IsTestnet, tt.wantTestnet)
			}
		})
	}
}

func TestGetNetworkConfigInvalid(t *testing.T) {
	tests := []string{
		"",
		"invalid",
		"eip155:1",
		"polkadot:invalid",
		"polkadot:",
	}

	for _, network := range tests {
		t.Run(network, func(t *testing.T) {
			_, ok := GetNetworkConfig(network)
			if ok {
				t.Errorf("GetNetworkConfig(%v) should return false for invalid network", network)
			}
		})
	}
}

func TestNetworksMapCompleteness(t *testing.T) {
	expectedNetworks := []string{
		PolkadotAssetHubCAIP2,
		KusamaAssetHubCAIP2,
		WestendAssetHubCAIP2,
	}

	for _, network := range expectedNetworks {
		if _, ok := Networks[network]; !ok {
			t.Errorf("Networks map missing %v", network)
		}
	}

	// Verify count
	if len(Networks) != len(expectedNetworks) {
		t.Errorf("Networks map has %v entries, want %v", len(Networks), len(expectedNetworks))
	}
}
